clients/unshaded/src/test/java/tachyon/client/file/FileInStreamTest.java

```diff
@@ -0,0 +1,331 @@
+import tachyon.client.block.BlockInStream;
+import tachyon.client.block.BufferedBlockInStream;
+import tachyon.client.block.TachyonBlockStore;
+import tachyon.client.block.TestBufferedBlockInStream;
+import tachyon.client.block.TestBufferedBlockOutStream;
+import tachyon.client.file.options.InStreamOptions;
+import tachyon.client.util.ClientMockUtils;
+import tachyon.client.util.ClientTestUtils;
+import tachyon.exception.ExceptionMessage;
+import tachyon.exception.PreconditionMessage;
+import tachyon.thrift.FileInfo;
+import tachyon.underfs.UnderFileSystem;
+import tachyon.util.io.BufferUtils;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({FileSystemContext.class, TachyonBlockStore.class, UnderFileSystem.class})
+public class FileInStreamTest {
+
+  private static final long BLOCK_LENGTH = 100L;
+  private static final long FILE_LENGTH = 350L;
+  private static final long NUM_STREAMS = ((FILE_LENGTH - 1) / BLOCK_LENGTH) + 1;
+
+  private TachyonBlockStore mBlockStore;
+  private FileSystemContext mContext;
+  private FileInfo mInfo;
+
+  private List<BufferedBlockInStream> mInStreams;
+  private List<TestBufferedBlockOutStream> mCacheStreams;
+
+  private FileInStream mTestStream;
+
+  @Rule
+  public final ExpectedException mThrown = ExpectedException.none();
+
+  @Before
+  public void before() throws IOException {
+    mInfo = new FileInfo().setBlockSizeBytes(BLOCK_LENGTH).setLength(FILE_LENGTH);
+
+    ClientTestUtils.setSmallBufferSizes();
+
+    mContext = PowerMockito.mock(FileSystemContext.class);
+    mBlockStore = PowerMockito.mock(TachyonBlockStore.class);
+    Mockito.when(mContext.getTachyonBlockStore()).thenReturn(mBlockStore);
+
+    // Set up BufferedBlockInStreams and caching streams
+    mInStreams = Lists.newArrayList();
+    mCacheStreams = Lists.newArrayList();
+    List<Long> blockIds = Lists.newArrayList();
+    for (int i = 0; i < NUM_STREAMS; i ++) {
+      blockIds.add((long) i);
+      mInStreams.add(new TestBufferedBlockInStream(i, (int) (i * BLOCK_LENGTH), BLOCK_LENGTH));
+      mCacheStreams.add(new TestBufferedBlockOutStream(i, BLOCK_LENGTH));
+      Mockito.when(mBlockStore.getInStream(i)).thenReturn(mInStreams.get(i));
+      Mockito
+          .when(
+              mBlockStore.getOutStream(Mockito.eq((long) i), Mockito.eq(-1L), Mockito.anyString()))
+          .thenReturn(mCacheStreams.get(i));
+    }
+    mInfo.setBlockIds(blockIds);
+
+    mTestStream = new FileInStream(mInfo, new InStreamOptions.Builder(ClientContext.getConf())
+        .setTachyonStorageType(TachyonStorageType.PROMOTE).build());
+    Whitebox.setInternalState(mTestStream, "mContext", mContext);
+  }
+
+  @After
+  public void after() {
+    ClientContext.reset();
+  }
+
+  @Test
+  public void singleByteReadTest() throws Exception {
+    // Verify byte by byte read is equal to increasing byte array
+    for (int i = 0; i < FILE_LENGTH; i ++) {
+      Assert.assertEquals((byte) i, (byte) mTestStream.read());
```

Comment:
Should this be like `Assert.assertEquals(i & 0xFF, mTestStream.read())`? If `mTestStream.read()` returns a value > 255 the cast to `byte` will hide that?